File: corpus/bench_qnorm.cpp
// C++ twin of bench_qnorm.lu — idiomatic style: AoS struct array, heap vector.
#include <cstdio>
#include <vector>

struct Quat {
  double w, x, y, z;
};

double qq(const std::vector<Quat>& qs, long n) {
  double s = 0.0;
  for (long i = 0; i < n; i++)
    s += qs[i].w * qs[i].w + qs[i].x * qs[i].x + qs[i].y * qs[i].y + qs[i].z * qs[i].z;
  return s;
}

int main() {
  const long n = 2000000;
  std::vector<Quat> qs(n);
  for (long i = 0; i < n; i++) {
    double f = i * 0.000001;
    qs[i] = Quat{1.0 + f, 2.0 - f, f * 3.0, 0.5};
  }
  double acc = 0.0;
  for (int r = 0; r < 20; r++) acc += qq(qs, n);
  printf("acc: %.17g\n", acc);
}

File: corpus/bench_qnorm.ts
// bench_qnorm.ts — TS twin of bench_qnorm.lu (typed arrays, SoA by hand:
// Bun's best case; idiomatic object arrays would be far slower).
const n = 2_000_000;
const w = new Float64Array(n);
const x = new Float64Array(n);
const y = new Float64Array(n);
const z = new Float64Array(n);
for (let i = 0; i < n; i++) {
  const f = i * 0.000001;
  w[i] = 1.0 + f;
  x[i] = 2.0 - f;
  y[i] = f * 3.0;
  z[i] = 0.5;
}

function qq(): number {
  let s = 0.0;
  for (let i = 0; i < n; i++) s += w[i] * w[i] + x[i] * x[i] + y[i] * y[i] + z[i] * z[i];
  return s;
}

let acc = 0.0;
for (let r = 0; r < 20; r++) acc += qq();
console.log("acc:", acc);
